### DISCUSSION TITLE
alittle confused about initialRemainingTime vs remainingTime

### DISCUSSION THREAD

**Discussion by loshya** [vydimitrov/react-countdown-circle-timer#79] [General]
i noticed an issue that the timer only starts when the window is focused so i tried to overcome this by using a date object from when the component was mounted vs the current time (which essentially would give me the true elapsed time. but the component always seems to rely on the remainingtime even when i try to change the key

any idea how to resolve this ?

**Comment by loshya** [comment]
quick example to maybe clear any confusion , i get sent something from the server at time x , i focus on the window ten seconds later 
so i want that to be reflected on the duration of the timer ( so if for example i had the duration set to 60 seconds but i open the tab ten seconds later i want it to display 50 seconds or 10 seconds have passed from the timer)

**Reply by vydimitrov** [reply]
In this case the `initialRemainingTime` should be 50, `duration ` should be 60

**Comment by vydimitrov** [comment]
Hey @loshya, can you provide a small demo where you reproduce the issue so I can take a look at it? Just by reading your comment I am not able to tell you what the problem is. 

Otherwise, your approach seems correct, when the component is mounted, you check how much time elapsed, then from the total time you remove the elapsed time and feed that to the component as `initialRemainingTime` (with a new `key` in case the component was mounted before).

**Reply by loshya** [reply]
Unfortunately i could not reproduce the issue as i was using very specific case where i had data come in from a websocket along with a timestamp , and i would check the difference between that timestamp and one from date.now() (when tab is focused)

**Reply by loshya** [reply]
the more frustrating issue is that i did console log the function which checks the elapsed time difference from both timestamps and it works but when i use it in initialRemainingTime , it considers the difference as zero and so the counter shows full duration rather than the one calculated from my function

**Reply by vydimitrov** [reply]
Can you share only the code for the timer where you feed the data? I may be able to spot something there.

**Reply by loshya** [reply]
`function test(){
return data.duration -  Math.floor((Date.now()-data.created_at) / 1000) 
} 
<CountdownCircleTimer
              isPlaying
              key={data.key}
              onComplete={() => {
               //handle data
              }}
              duration={data.duration}
              initialRemainingTime={test()}
              size={80}
              strokeWidth={7}
              colors="#A30000"
            >
              {({ remainingTime }) => {
                console.log(remainingTime,test())
                return remainingTime;
              }}
            </CountdownCircleTimer>
`
im not sure if this helps but maybe that would show something

**Reply by vydimitrov** [reply]
Is this how your App looks?
https://codesandbox.io/s/ecstatic-nash-o6riy?file=/src/index.js

**Reply by loshya** [reply]
that is correct
if you try and open the app in a new window and check the console you will see something similar to this if you wait a few seconds before focusing/ opening the tab 

recreated here - https://imgur.com/a/k35kI8w

**Reply by vydimitrov** [reply]
In my example, I do not change the key so that will not reset the time. Can you record a video and posted here. I am not sure what I am looking in the image you sent to me.

**Reply by loshya** [reply]
here i refresh the tab and go into another tab before it loads and wait around 10 seconds before reponening the app window
you can see in the console that the log from remainingTime and the function is different 

https://imgur.com/a/6GjIG9I

**Reply by vydimitrov** [reply]
I see what you mean .... so what happens here is that when you init the app you init the Timer with some duration and `initialRemaingTime`. If you navigate away and you come back to the tab the timer will continue using the old values. By design the duration and `initialRemaingTime` can not be just changed. Read more about it [here](https://github.com/vydimitrov/react-countdown-circle-timer#change-duration-andor-initialremainingtime-once-component-is-mounted).

What you need is a listener for when the tab becomes active again to update the `key` prop of the Timer, which will reset the duration and `initialRemaingTime` . One way of doing it is like this
```js
 // Active
 window.addEventListener('focus', //change the `key` of the Timer);

 // Inactive
 window.addEventListener('blur', stopTimer);
```

**Reply by vydimitrov** [reply]
You can use this one as well to pause and play the timer - https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event

**Reply by vydimitrov** [reply]
Here is a working demo where you pause the timer when the tab is not focused - https://codesandbox.io/s/gifted-murdock-ly719?file=/src/index.js

Does this help?

**Comment by ghost** [comment]
I have this same issue. Countdown starts only when tab is active. It should start tho even if I am not in the tab right now.. Anyone has a fix for this?
